### ISSUE TITLE
Wish I could get the record key in async producer send callback

### ISSUE THREAD

**Issue by berrfred** [dpkp/kafka-python#2452]
Hi, while producing records asynchronously, I'd like to be able to log the provided key together with topic/partition/offset upon successfull send ... if that makes sense to you.

```
#--- Print sent message partition and offset upon success
def on_send_success(record_metadata):
	#clog.log_(LOG_WARNING, f"sent|{record_metadata.topic}|{record_metadata.partition}|{record_metadata.offset}|{record_metadata.key}".encode('utf-8'))
	clog.log_(LOG_WARNING, f"sent|{record_metadata}".encode('utf-8'))
```

Today only the key size is available but not its value.

16/12/2024; 13:43:00.11; [W] $GDRK0 ; sent|RecordMetadata(**topic**='oamds-dr', **partition**=0, topic_partition=TopicPartition(topic='oamds-dr', partition=0), **offset**=131853, timestamp=1734352978359, log_start_offset=131838, checksum=None, **serialized_key_size**=44, serialized_value_size=55, serialized_header_size=66)

**Comment by zxsa89675** [comment]
Hello, we tried to solve the issue.

This is what we did:

Modify the RecordMetadata namedtuple to include the 'key' field and update the FutureRecordMetadata class to pass the key through to the RecordMetadata. This change allows access to the key value when logging successful sends.

You can review changes in this commit: https://github.com/zxsa89675/dpkp-kafka-python-2452/commit/1c81b01252eca7d9cab5c52bea0e0119ef7c6a22.


> [!CAUTION]
> Disclaimer: The concept of solution was created by AI and you should never copy paste this code before you check the correctness of generated code. Solution might not be complete, you should use this code as an inspiration only.

---

If you no longer want Latta AI to attempt solving issues on your repository, you can block this account.

**Comment by dpkp** [comment]
You have the key when you get the future and add the callback, so you should be able to add it as an extra arg:

```
producer.send(..., key=my_key,...).add_callback(on_send_success, my_key)

def

**Comment by berrfred** [comment]
Good point, I totally missed this possibility ... so passing key=my_key on the add_callback and managing it on the callback function does the job.  Thanks.
